src/Main.hs

```diff
@@ -147,6 +174,73 @@ checkSession db = do
         Just session -> return session
         Nothing -> raiseStatus status401 "Invalid session cookie"
     Nothing -> raiseStatus status401 "Invalid session cookie"
+data Ingredient = Ingredient {count :: Int, ingredient :: String} deriving (Read, Show, Generic)
+
+instance ToJSON Ingredient
+
+instance FromJSON Ingredient
+
+data Recipe = Recipe
+  { title :: String,
+    description :: String,
+    ingredients :: [Ingredient],
+    timeMinutes :: Int,
+    costCents :: Int,
+    stars :: Int
+  }
+  deriving (Generic)
+
+instance ToJSON Recipe
+
+instance FromJSON Recipe
+
+instance FromRow Recipe where
+  fromRow = Recipe <$> field <*> field <*> (read <$> field) <*> field <*> field <*> field
+
+instance ToRow Recipe where
+  toRow (Recipe t d i tM cC s) =
+    [ toField t,
+      toField d,
+      toField . show $ i,
+      toField tM,
+      toField cC,
+      toField s
+    ]
+
+type RecipeID = Int
+
+data RecipeWithID = RecipeWithID {recipeID :: Int, recipe :: Recipe}
+
+instance FromRow RecipeWithID where
+  fromRow = RecipeWithID <$> field <*> fromRow
+
+instance ToJSON RecipeWithID where
+  toJSON (RecipeWithID rID (Recipe t d i tM cC s)) =
+    object
+      [ "id" .= rID,
+        "title" .= t,
+        "description" .= d,
+        "ingredients" .= i,
+        "timeMinutes" .= tM,
+        "costCents" .= cC,
+        "stars" .= s
+      ]
+
+getAllRecipes :: Connection -> IO [RecipeWithID]
+getAllRecipes = flip query_ "SELECT * FROM recipes"
```

Comment:
semicolon in query fehlt

Reply:
Resolved in 9d8ac164eaaf6e8355e1254ae071ac5eb29b6aef